x-pack/filebeat/input/s3/s3_integration_test.go

```diff
@@ -150,103 +173,68 @@ func defaultTestConfig() *common.Config {
-	defer eventOutlet.Close()
-
-	connector := channel.ConnectorFunc(func(_ *common.Config, _ beat.ClientConfig) (channel.Outleter, error) {
-		return eventOutlet, nil
-	})
-
-	in, err := NewInput(cfg, connector, inputCtx)
+func runTest(t *testing.T, cfg *common.Config, run func(t *testing.T, input *s3Collector, receiver *eventReceiver)) {
+	plugin := Plugin()
+	inp, err := plugin.Manager.Create(cfg)
 	if err != nil {
 		t.Fatal(err)
 	}
-	s3Input := in.(*s3Input)
-	defer s3Input.Stop()
 
-	run(t, s3Input, eventOutlet)
-}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
 
-func newInputContext() input.Context {
-	return input.Context{
-		Done: make(chan struct{}),
+	client := pubtest.NewChanClient(0)
+	receiver := &eventReceiver{client.Channel}
+	pipeline := pubtest.ConstClient(client)
+	collector, err := inp.(*s3Input).createCollector(v2.Context{
+		Logger:      logp.NewLogger("test"),
+		Cancelation: ctx,
+	}, pipeline)
+	if err != nil {
+		t.Fatal(err)
 	}
-}
 
-type stubOutleter struct {
-	sync.Mutex
-	cond   *sync.Cond
-	done   bool
-	Events []beat.Event
+	run(t, collector, receiver)
 }
 
-func newStubOutlet() *stubOutleter {
-	o := &stubOutleter{}
-	o.cond = sync.NewCond(o)
-	return o
+type eventReceiver struct {
+	ch chan beat.Event
 }
 
-func (o *stubOutleter) waitForEvents(numEvents int) ([]beat.Event, bool) {
-	o.Lock()
-	defer o.Unlock()
-
-	for len(o.Events) < numEvents && !o.done {
-		o.cond.Wait()
-	}
-
-	size := numEvents
-	if size >= len(o.Events) {
-		size = len(o.Events)
+func (r *eventReceiver) waitForEvents(n int) ([]beat.Event, bool) {
+	buf := make([]beat.Event, 0, n)
+	for event := range r.ch {
+		buf = append(buf, event)
+		n--
+		if n == 0 {
+			return buf, true
+		}
 	}
-
-	out := make([]beat.Event, size)
-	copy(out, o.Events)
-	return out, len(out) == numEvents
-}
-
-func (o *stubOutleter) Close() error {
-	o.Lock()
-	defer o.Unlock()
-	o.done = true
-	return nil
-}
-
-func (o *stubOutleter) Done() <-chan struct{} { return nil }
-
-func (o *stubOutleter) OnEvent(event beat.Event) bool {
-	o.Lock()
-	defer o.Unlock()
-	o.Events = append(o.Events, event)
-	o.cond.Broadcast()
-	return !o.done
+	return buf, false
 }
 
 func TestS3Input(t *testing.T) {
 	inputConfig := defaultTestConfig()
 	config := getConfigForTest(t)
 
-	runTest(t, inputConfig, func(t *testing.T, input *s3Input, out *stubOutleter) {
+	runTest(t, inputConfig, func(t *testing.T, input *s3Collector, receiver *eventReceiver) {
 		awsConfig, err := awscommon.GetAWSCredentials(config.AwsConfig)
 		if err != nil {
-
+			t.Fatal("failed GetAWSCredentials with AWS Config: ", err)
 		}
+
 		s3BucketRegion := os.Getenv("S3_BUCKET_REGION")
 		if s3BucketRegion == "" {
 			t.Log("S3_BUCKET_REGION is not set, default to us-west-1")
 			s3BucketRegion = "us-west-1"
 		}
 		awsConfig.Region = s3BucketRegion
-		input.awsConfig = awsConfig.Copy()
+		awsConfig = awsConfig.Copy()
 		svcSQS := sqs.New(awsConfig)
+		input.sqs = svcSQS
```

Comment:
Having to modify the object under test for the test is kind of a red flag to me. It should be possible to create and provide a fully initialized object. Is there something missing in the construction of the collector that forces us to do this here?

Reply:
@urso I moved the code for creating a full initialized object outside of `runTest` function. Please let me know if this looks less suspicious for you now :)